### ISSUE TITLE
MIT/SIT Strategy unification

### ISSUE THREAD

**Issue by AntonioCarta** [ContinualAI/avalanche#279]
I'm opening this issue to propose the unification of SIT/MT/MIT scenarios into a single `BaseStrategy`.
**Why**: Currently, most plugins should support MIT. However, if we implement them differently, the result is that we need multiple versions of the same plugin. Supporting MIT scenarios directly seems easier to me.

The basic loop does not change much:
```
train:
	before_training
	train_step:					 			
		adapt_train_dataset						
		make_train_dataloader                   <-- dataloader SIT/MT
		before_training_step
		for epoch:
			before_training_epoch
			training_epoch:						<-- loop t, <x, y>
				before_training_iteration
				
				before_forward
				batch_forward:					<-- loop over tasks
					for task:
						model.forward
				after_forward
				
				before_backward
				after_backward
				
				before_update
				after_update
				
				after_training_iteration
			after_training_epoch
	after_training
```
Test is similar.

As you can see, the meaningful changes are:
1 - the dataloader (think about the current MultiTaskDataLoader)
2 - the data format t, <x, y> instead of <x, y>
3 - the loop over tasks summing the loss over multiple tasks (like JointTraining)

(2) requires a lot of small changes, therefore I want to make sure to make it right. Notice that with this change the CL scenario (SIT/MT/MIT) depends only on the data.

**Data Stream**: I'm assuming that each step in the stream has two labels <step_id, task_id>. The first one is private, i.e. it must not be used by strategies. The second one can be used. Therefore a scenario is SIT if task_labels=False.

What do you think?
